### PR TITLE
Finalization of the ePayments V1 API

### PR DESCRIPTION
**Pull request by tht13**

### CODE REVIEW

File: docs/epayment.v1.yml [hunk 1]
```diff
@@ -246,7 +210,10 @@ components:
           pattern: '^[0-9-]{10,15}$'
         sub:
           type: string
-          description: The user subscription identifier received in `$.profile.sub` or from the Vipps Login API
```
Comment (cloveras) [inline docs/epayment.v1.yml:249]: Change to simply `sub`?

File: docs/epayment.v1.yml [hunk 2]
```diff
@@ -28,7 +27,6 @@ tags:
 paths:
   /v1/payments:
     parameters:
-      - $ref: '#/components/parameters/Idempotency-Key'
```
Comment (cloveras) [inline docs/epayment.v1.yml:31]: Why no idempotency key?

**Comment by cloveras** [comment]
Looks good!

I only have some minor comments, mostly just prettification of the YAML:
- editor.swagger.io complains: should be an array of property names required within an object schema: Jump to line 261
- Some object lacks a description, like "PaymentMethod".
- Try to add examples everywhere. For instance for "scope". See standard examples here: https://github.com/vippsas/vipps-rfc/blob/master/docs/0001-api-guidelines.md#examples-of-data-types-and-values
- Explicit examples, like "https://example.com/vipps/callbacks-for-payment-update-from-vipps" for callbackUrl, reduces confusion
- There are some occurrences of non-standard  "x-whatnot" (from the IDE/editor, I guess) that can be removed.
- Whitespace in "Payment Response", unlike "PaymentRequest". Best without.
-

File: docs/epayment.v1.yml [hunk 3]
```diff
@@ -344,18 +293,57 @@ components:
...
           type: string
-          description: 'message to user that will appear in the app, in the merchant portal and settlement files.'
+          minLength: 3
+          maxLength: 100
+          description: 'The message to user that will appear in the app, in the merchant portal and settlement files.'
         callbackUrl:
           type: string
-          description: 'This is an URL for Vipps to call at the merchant''s server to provide updated information about the order after the payment request. Domain name and context path should be provided by merchant as the value for this parameter. Vipps will add `/payments/{reference}` to the end of this URL.'
+          description: |
+            TODO add regex validation
```
Comment (khenrichsen) [inline docs/epayment.v1.yml:304]: hehe, was just about to say - add regex validation here

**Review by khenrichsen — COMMENTED** [review]
Additional comments (on lines i cant comment on this PR):

airlineCode line 360 - why string?

Payment-Response on initiate? Maybe add two seperate examples for the merchants what to expect on each

logisticsOptions: line 525
add min and max value to all fields

image
swagger error on customerInteraction

File: docs/epayment.v1.yml [hunk 4]
```diff
@@ -299,36 +258,26 @@ components:
...
-          $ref: '#/components/schemas/OrderInformation'
+        receipt:
+          $ref: 'https://raw.githubusercontent.com/vippsas/vipps-order-management-api/2.3.0/docs/swagger.yaml#/components/schemas/Receipt'
         paymentMethod:
           $ref: '#/components/schemas/PaymentMethod'
         profile:
-          $ref: '#/components/schemas/Profile'
-        pspReference:
-          $ref: '#/components/schemas/PspReferenceType'
-        redirectUrl:
```
[the remark this answers is not in the document]
Comment (khenrichsen) [reply docs/epayment.v1.yml:319]: if the "end-state" of the payment should be allowed to be controlled by the merchant - e.g. RECEIPT, HOME_PAGE, WEB_REDIRECT ++

[the remark this answers is not in the document]
Comment (khenrichsen) [reply docs/epayment.v1.yml:319]: and it deffo needs validation

File: docs/epayment.v1.yml (hunk 2 above)
Comment (tht13) [reply docs/epayment.v1.yml:31]: re-adding, there are actually some future usecases that it would be good to have

**Pull request closed without merging**